### ISSUE TITLE
why upsample? 

### ISSUE THREAD

**Issue by Yuheng-Li** [NVlabs/stylegan3#91]
Why we need upsapmpling? what would happen if we do: activation and apply low pass filters corresponding to output of that layer?

**Comment by PeterL1n** [comment]
The reason is that we want to interpret the feature as a discretized version of a continuous signal. When you do activation directly on the discretized features, it introduces aliasing. Sure you can apply low pass filter after it, which acts as a blur filter to reduce aliasing. But if the blur is low amount, the aliasing is still present. If the blur is high, then you are essentially reducing the frequency of the signal by much. Which is wasting the resolution of the feature map. Nvidia's approach is to only upsample and downsample for the activation. They implement it in a custom kernel which is surprisingly efficient.

**Comment by Yuheng-Li** [comment]
I agree that we need to carefully design low pass filter. But if I do not do upsampling and directly apply the same filter designed by "out_sampling_rate", "out_cutoff" and "out_half_width" after activation, what is the difference does it make here?

**Comment by PeterL1n** [comment]
So for a concrete example. Let's say your feature map is 32x32. After convolution, you apply leakyrelu activation. Now the frequency of the signal becomes fs=16. Note that you now introduced aliasing. [See Video](https://nvlabs-fi-cdn.nvidia.com/_web/stylegan3/videos/video_7_figure_2_right_filtered_nonlinearity.mp4). If you then imagine applying an ideal sinc filter with fc=16, it really should not do anything at all. But since ideal filter is not really possible, we used kaiser-windowed sinc filter. The kaiser-windowed filter is an approximation so it will slightly blur the feature map, but that is not sufficient to remove the aliasing.

What the paper did was to 2x upsample the feature map to 64x64. The input signal after convolution still has frequency fs=16. Then you apply activation, which makes frequency fs=32. Then we cut the frequency back to fs=16 with kaiser sinc. And downsample the feature map back to 32x32. The author experimented with different upsample factor (2x, 4x, 8x) and found 2x is sufficient.

**Comment by betterze** [comment]
@PeterL1n The SG3 perform upsampling by interleaving the feature map with 0 then convolution with the upsampling filter. The function for upsampling is in [here](https://github.com/NVlabs/stylegan3/blob/main/torch_utils/ops/filtered_lrelu.py#L146). 

Could you explain why the network perform upsampling in this way? The design for [upsampling filter](https://github.com/NVlabs/stylegan3/blob/main/training/networks_stylegan3.py#L310) is very complex. I have little signal processing background and I am not able to understand it. 

I try to visualize the 'x' before and after this [function](https://github.com/NVlabs/stylegan3/blob/main/torch_utils/ops/filtered_lrelu.py#L146). The images look similar, but the values do not seem to follow simple bilinear interpolation upsampling. 

Is there a way I can invert this upsampling operation given a specific upsampling filter? 

Thank you for your advice.
